src/hooks/useEditorView.ts

```diff
@@ -59,70 +36,86 @@ function withFlushedDispatch(
  * All state and props updates are executed in a layout effect.
  * To ensure that the EditorState and EditorView are never out of
  * sync, it's important that the EditorView produced by this hook
- * is only accessed through the `useEditorViewEvent` and
- * `useEditorViewLayoutEffect` hooks.
+ * is only accessed through the provided hooks.
  */
 export function useEditorView<T extends HTMLElement = HTMLElement>(
   mount: T | null,
-  props: EditorProps
-): EditorView | null {
-  const [view, setView] = useState<EditorView | null>(null);
-  const forceUpdate = useForceUpdate();
-
-  const editorProps = withFlushedDispatch(props, forceUpdate);
+  props: Props
+): EditorContextValue {
+  if (process.env.NODE_ENV !== "production") {
+    if (
+      props.defaultState !== undefined &&
+      props.state !== undefined &&
+      !didWarnValueDefaultValue
+    ) {
+      console.error(
+        "A component contains a ProseMirror editor with both value and defaultValue props. " +
+          "ProseMirror editors must be either controlled or uncontrolled " +
+          "(specify either the value prop, or the defaultValue prop, but not both). " +
+          "Decide between using a controlled or uncontrolled ProseMirror editor " +
+          "and remove one of these props. More info: " +
+          "https://reactjs.org/link/controlled-components"
+      );
+      didWarnValueDefaultValue = true;
+    }
+  }
 
-  const stateProp = "state" in editorProps ? editorProps.state : undefined;
+  const defaultState = props.defaultState ?? EMPTY_STATE;
+  const [_state, setState] = useState<EditorState>(defaultState);
+  const state = props.state ?? _state;
 
-  const state =
-    "defaultState" in editorProps
-      ? editorProps.defaultState
-      : editorProps.state;
+  const {
+    componentEventListenersPlugin,
+    registerEventListener,
+    unregisterEventListener,
+  } = useComponentEventListeners();
 
-  const nonStateProps = Object.fromEntries(
-    Object.entries(editorProps).filter(
-      ([propName]) => propName !== "state" && propName !== "defaultState"
-    )
+  const plugins = useMemo(
+    () => [...(props.plugins ?? []), componentEventListenersPlugin],
+    [props.plugins, componentEventListenersPlugin]
   );
 
-  useLayoutEffect(() => {
-    return () => {
-      if (view) {
-        view.destroy();
+  function dispatchTransaction(this: EditorView, tr: Transaction) {
+    flushSync(() => {
+      if (props.dispatchTransaction) {
+        props.dispatchTransaction.call(this, tr);
+      } else {
+        setState((s) => s.apply(tr));
       }
```

Comment:
If I'm reading this correctly, it's making the assumption that users won't pass `dispatchTransaction` and `defaultState`, but that is a supported use case now, and I think it should continue to be. React allows this, for example, for uncontrolled components; it allows the consumer to execute side effects when changes occur, even if they haven't lifted the state.

Reply:
Yup. I think we could maybe make a deviation from EditorView here and stop passing the view as the `this` context so that it's not tempting to write `this.updateState()` in the transaction handler.

You can observe transactions, if you want to for some reason, but unless you control the state you shouldn't actually see the state in that callback or set it.

Reply:
I'm not sure I understand why. I mean I do think calling `this.updateState()` from a `dispatchTransaction` prop is goofy and maybe worth avoiding (though I think it was handled more or less fine with the `forceUpdate()`), but `<input />` `onChange` handlers get an event that allows you retrieve the entire full state of the element.

Reply:
Would you feel alright about pulling this out into a separate change? I think I would like to think about and discuss it more, but don't want to block this fix

Reply:
What do you want me to pull out? I haven't changed the signature of `dispatchTransaction` here, I'm just suggesting we consider it.

Reply:
Ah, this is more subtle than I remembered/was envisioning. Sorry, I could have been being clearer about this the whole time. The `forceUpdate()` from before would handle the case where someone called `this.updateState()` from `dispatchTransaction`. Storing the editor state in React state here when the component is uncontrolled is definitely the right move, but I think if there's a dispatchTransaction _and_ a defaultState right now, we wouldn't re-render the component tree. So what I'm asking to be backed out is that regression!

Reply:
Oh, I see! I have to think on that. I think that's pretty central to this PR, so rather than back out I think we need to address it.

Handling this correctly might really require intercepting calls to `updateState` or `setProps`. I dunno.

I'm almost tempted to just rip out support for uncontrolled state at this point.

Reply:
> I think that's pretty central to this PR

I don't understand why that would be; couldn't we do a forceUpdate() after calling `props.dispatchTransaction()` if `props.state` is undefined?

Reply:
I guess we could force an update, but as long as we're storing the uncontrolled state in a state hook, there's no point. That state hasn't changed. If there's other reactive state that the use changed with their `dispatchTransaction`, then that'll cause a render.

I could see calling `setState` before calling the user's `dispatchTransaction` so that, much like regular inputs, the view instance already has an updated value at that point and, like `onChange` you could observe it. In the case of `dispatchTransaction` that's through `this.state`. In the case of `onChange`, `event.target.value`.

I don't know that we should do it when state is controlled, unless we want folks to just observe `this.state` and set it somewhere in the common case. Internally, in the Oak application, we generally apply the transaction ourselves in Redux reducers, such that applying it in the hook is duplicative work.

If we wanna get fancy, we could call `dispatchTransaction` with some kind of proxy object as `this`.

Reply:
I updated this to call `setState()` before calling `props.dispatchTransaction()` if `props.state` is not specified.

I think that should cause any re-render triggered by `dispatchTransaction` higher up in the tree to flush the uncontrolled state change to the view, doesn't cause any redundant work for the case of controlled state, and ensures that transactions get applied even when `dispatchTransaction` and `defaultState` get specified together.

If you don't try to control the state, it'll be semi-controlled internally, in the sense that we keep state in the hook. A `dispatchTransaction` prop won't be able to observe the state having been applied to the actual view instance yet, but I think maybe this is good enough.

Reply:
I've now added a test for this case, too.

Reply:
Let me know if I captured and addressed your concerns correctly, here.